pkg/manager/tidbngmonitoring/ng_monitoring_manager_test.go

```diff
@@ -0,0 +1,850 @@
+
+			err := manager.syncService(tngm)
+			testcase.expectErr(err)
+		}
+	})
+
+	t.Run("syncCore", func(t *testing.T) {
+		g := NewGomegaWithT(t)
+
+		type testcase struct {
+			name string
+
+			setTNGM              func(tngm *v1alpha1.TidbNGMonitoring)
+			getSts               func() (*apps.StatefulSet, *apps.StatefulSet) // old sts is nil means that sts isn't found
+			populateStatusErr    error
+			syncConfigMapErr     error
+			createStatefulSetErr error
+			updateStatefulSetErr error
+			expectFn             func(tngm *v1alpha1.TidbNGMonitoring, err error)
+		}
+
+		cases := []testcase{
+			{
+				name: "populate status failed",
+				getSts: func() (*apps.StatefulSet, *apps.StatefulSet) {
+					return nil, nil
+				},
+				populateStatusErr:    fmt.Errorf("populate status failed"),
+				syncConfigMapErr:     fmt.Errorf("shouldn't sync configmap"),
+				createStatefulSetErr: fmt.Errorf("shouldn't create sts"),
+				updateStatefulSetErr: fmt.Errorf("shouldn't update sts"),
+				expectFn: func(tngm *v1alpha1.TidbNGMonitoring, err error) {
+					g.Expect(err).Should(HaveOccurred())
+					g.Expect(err.Error()).Should(ContainSubstring("populate status failed"))
+				},
+			},
+			{
+				name: "manager is paused",
+				setTNGM: func(tngm *v1alpha1.TidbNGMonitoring) {
+					tngm.Spec.Paused = true
+				},
+				getSts: func() (*apps.StatefulSet, *apps.StatefulSet) {
+					return nil, nil
+				},
+				syncConfigMapErr:     fmt.Errorf("shouldn't sync configmap"),
+				createStatefulSetErr: fmt.Errorf("shouldn't create sts"),
+				updateStatefulSetErr: fmt.Errorf("shouldn't update sts"),
+				expectFn: func(tngm *v1alpha1.TidbNGMonitoring, err error) {
+					g.Expect(err).Should(Succeed())
+				},
+			},
+			{
+				name: "sync configmap failed",
+				getSts: func() (*apps.StatefulSet, *apps.StatefulSet) {
+					return nil, nil
+				},
+				syncConfigMapErr:     fmt.Errorf("should sync configmap"),
+				createStatefulSetErr: fmt.Errorf("shouldn't create sts"),
+				updateStatefulSetErr: fmt.Errorf("shouldn't update sts"),
+				expectFn: func(tngm *v1alpha1.TidbNGMonitoring, err error) {
+					g.Expect(err).Should(HaveOccurred())
+					g.Expect(err.Error()).Should(ContainSubstring("should sync configmap"))
+				},
+			},
+			{
+				name: "should create sts if old sts isn't found",
+				getSts: func() (*apps.StatefulSet, *apps.StatefulSet) {
+					sts := &apps.StatefulSet{}
+					sts.Name = NGMonitoringName("ngm")
+					sts.Namespace = "default"
+					return nil, sts.DeepCopy()
+				},
+				createStatefulSetErr: fmt.Errorf("should create sts"),
+				updateStatefulSetErr: fmt.Errorf("shouldn't update sts"),
+				expectFn: func(tngm *v1alpha1.TidbNGMonitoring, err error) {
+					g.Expect(err).Should(HaveOccurred())
+					g.Expect(err.Error()).Should(ContainSubstring("should create sts"))
+				},
+			},
+			{
+				name: "should update sts if old sts is exist",
```

Comment:
This case check if call the `UpdateSevice`

Reply:
equal svc will not trigger an update, but equal sts will trigger an update?

Reply:
No.
The func `UpdateStatefulSet` checks if sts is equal, it doesn't matter for the case. The case just ensure it will call `UpdateStatefulSet` when old sts is exist